### ISSUE TITLE
training errors using preprocessed data

### ISSUE THREAD

**Issue by gabrielepozzati** [baldassarreFe/protein-quality-gn#3]
I followed the readme instructions to get the miniconda environment and got the preprocessed data. 
After running into the environment:

python -m src.graphqa.train config/train.yaml

I get the error: 

```
Validation sanity check: 100%|██████████| 1/1.0 [00:00<00:00,  1.88it/s]2020-11-18 09:23:26.893 | ERROR    | __main__:<module>:299 - An error has been caught in function '<module>', process 'MainProcess' (16505), thread 'MainThread' (140044919113536):
Traceback (most recent call last):

  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
           │         │     └ {'__name__': '__main__', '__doc__': None, '__package__': 'src.graphqa', '__loader__': <_frozen_importlib_external.SourceFileL...
           │         └ <code object <module> at 0x7f5ebf05db30, file "/storage/graphqa/src/graphqa/train.py", line 1>
           └ <function _run_code at 0x7f5ebf03bee0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
         │     └ {'__name__': '__main__', '__doc__': None, '__package__': 'src.graphqa', '__loader__': <_frozen_importlib_external.SourceFileL...
         └ <code object <module> at 0x7f5ebf05db30, file "/storage/graphqa/src/graphqa/train.py", line 1>

> File "/storage/graphqa/src/graphqa/train.py", line 299, in <module>
    main()
    └ <function main at 0x7f5e25a05c10>

  File "/storage/graphqa/src/graphqa/train.py", line 292, in main
    result = trainer.fit(model)
             │       │   └ LightningGraphQA(
             │       │       (model): RecursiveScriptModule(
             │       │         original_name=GraphQA
             │       │         (embeddings): RecursiveScriptModule(
             │       │           ...
             │       └ <function Trainer.fit at 0x7f5e5c5ff280>
             └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>

  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/states.py", line 48, in wrapped_fn
    result = fn(self, *args, **kwargs)
             │  │      │       └ {}
             │  │      └ (LightningGraphQA(
             │  │          (model): RecursiveScriptModule(
             │  │            original_name=GraphQA
             │  │            (embeddings): RecursiveScriptModule(
             │  │             ...
             │  └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
             └ <function Trainer.fit at 0x7f5e5c5ff1f0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1073, in fit
    results = self.accelerator_backend.train(model)
              │    │                   │     └ LightningGraphQA(
              │    │                   │         (model): RecursiveScriptModule(
              │    │                   │           original_name=GraphQA
              │    │                   │           (embeddings): RecursiveScriptModule(
              │    │                   │             ...
              │    │                   └ <function GPUBackend.train at 0x7f5e6123be50>
              │    └ <pytorch_lightning.accelerators.gpu_backend.GPUBackend object at 0x7f5e25922940>
              └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/accelerators/gpu_backend.py", line 51, in train
    results = self.trainer.run_pretrain_routine(model)
              │    │       │                    └ LightningGraphQA(
              │    │       │                        (model): RecursiveScriptModule(
              │    │       │                          original_name=GraphQA
              │    │       │                          (embeddings): RecursiveScriptModule(
              │    │       │                            ...
              │    │       └ <function Trainer.run_pretrain_routine at 0x7f5e5c5ff430>
              │    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
              └ <pytorch_lightning.accelerators.gpu_backend.GPUBackend object at 0x7f5e25922940>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1224, in run_pretrain_routine
    self._run_sanity_check(ref_model, model)
    │    │                 │          └ LightningGraphQA(
    │    │                 │              (model): RecursiveScriptModule(
    │    │                 │                original_name=GraphQA
    │    │                 │                (embeddings): RecursiveScriptModule(
    │    │                 │                  ...
    │    │                 └ LightningGraphQA(
    │    │                     (model): RecursiveScriptModule(
    │    │                       original_name=GraphQA
    │    │                       (embeddings): RecursiveScriptModule(
    │    │                         ...
    │    └ <function Trainer._run_sanity_check at 0x7f5e5c5ff4c0>
    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1257, in _run_sanity_check
    eval_results = self._evaluate(model, self.val_dataloaders, max_batches, False)
                   │    │         │      │    │                └ [1.0]
                   │    │         │      │    └ [<torch_geometric.data.dataloader.DataLoader object at 0x7f5e176ee2b0>]
                   │    │         │      └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
                   │    │         └ LightningGraphQA(
                   │    │             (model): RecursiveScriptModule(
                   │    │               original_name=GraphQA
                   │    │               (embeddings): RecursiveScriptModule(
                   │    │                 ...
                   │    └ <function TrainerEvaluationLoopMixin._evaluate at 0x7f5e5c649310>
                   └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 413, in _evaluate
    self.__call_eval_loop_hook_end(test_mode)
    │                              └ False
    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 237, in __call_eval_loop_hook_end
    self.__call_eval_loop_hook_evt(test_mode, 'end')
    │                              └ False
    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 247, in __call_eval_loop_hook_evt
    getattr(self, hook_name)()
            │     └ 'on_validation_epoch_end'
            └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/callback_hook.py", line 86, in on_validation_epoch_end
    callback.on_validation_epoch_end(self, self.get_model())
    │        │                       │     │    └ <function TrainerIOMixin.get_model at 0x7f5e5c5d8f70>
    │        │                       │     └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
    │        │                       └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
    │        └ <function EarlyStopping.on_validation_epoch_end at 0x7f5e6128b5e0>
    └ <src.graphqa.callbacks.MyEarlyStopping object at 0x7f5e177f1910>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/callbacks/early_stopping.py", line 156, in on_validation_epoch_end
    self._validate_condition_metric(trainer.callback_metrics)
    │    │                          │       └ {'global': score         cad      gdtha     gdtts      lddt        tm
    │    │                          │         metric                                                 ...
    │    │                          └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f5e177f1790>
    │    └ <function EarlyStopping._validate_condition_metric at 0x7f5e6128b310>
    └ <src.graphqa.callbacks.MyEarlyStopping object at 0x7f5e177f1910>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/callbacks/early_stopping.py", line 117, in _validate_condition_metric
    raise RuntimeError(error_msg)
                       └ 'Early stopping conditioned on metric `val/global_per_target/gdtts/R` which is not available. Either add `val/global_per_targ...

RuntimeError: Early stopping conditioned on metric `val/global_per_target/gdtts/R` which is not available. Either add `val/global_per_target/gdtts/R` to the return of  validation_epoch end or modify your EarlyStopping callback to use any of the following: `global`, `local`, `global_per_target`, `local_per_decoy`, `global/gdtts`, `global/cad`, `local/lddt`, `local/cad`
```

So I tried to substitute "val/global_per_target/gdtts/R" with "global/gdtts" in config/train.yaml but then I get: 

```
Epoch 0:   0%|          | 1/694 [00:05<1:00:18,  5.22s/it, loss=0.550]2020-11-18 09:34:34.125 | ERROR    | __main__:<module>:299 - An error has been caught in function '<module>', process 'MainProcess' (16806), thread 'MainThread' (140263578617664):
Traceback (most recent call last):

  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
           │         │     └ {'__name__': '__main__', '__doc__': None, '__package__': 'src.graphqa', '__loader__': <_frozen_importlib_external.SourceFileL...
           │         └ <code object <module> at 0x7f91a8251b30, file "/storage/graphqa/src/graphqa/train.py", line 1>
           └ <function _run_code at 0x7f91a822fee0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
         │     └ {'__name__': '__main__', '__doc__': None, '__package__': 'src.graphqa', '__loader__': <_frozen_importlib_external.SourceFileL...
         └ <code object <module> at 0x7f91a8251b30, file "/storage/graphqa/src/graphqa/train.py", line 1>

> File "/storage/graphqa/src/graphqa/train.py", line 299, in <module>
    main()
    └ <function main at 0x7f910ebf9c10>

  File "/storage/graphqa/src/graphqa/train.py", line 292, in main
    result = trainer.fit(model)
             │       │   └ LightningGraphQA(
             │       │       (model): RecursiveScriptModule(
             │       │         original_name=GraphQA
             │       │         (embeddings): RecursiveScriptModule(
             │       │           ...
             │       └ <function Trainer.fit at 0x7f91457f3280>
             └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>

  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/states.py", line 48, in wrapped_fn
    result = fn(self, *args, **kwargs)
             │  │      │       └ {}
             │  │      └ (LightningGraphQA(
             │  │          (model): RecursiveScriptModule(
             │  │            original_name=GraphQA
             │  │            (embeddings): RecursiveScriptModule(
             │  │             ...
             │  └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
             └ <function Trainer.fit at 0x7f91457f31f0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1073, in fit
    results = self.accelerator_backend.train(model)
              │    │                   │     └ LightningGraphQA(
              │    │                   │         (model): RecursiveScriptModule(
              │    │                   │           original_name=GraphQA
              │    │                   │           (embeddings): RecursiveScriptModule(
              │    │                   │             ...
              │    │                   └ <function GPUBackend.train at 0x7f914a42ee50>
              │    └ <pytorch_lightning.accelerators.gpu_backend.GPUBackend object at 0x7f910eb19940>
              └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/accelerators/gpu_backend.py", line 51, in train
    results = self.trainer.run_pretrain_routine(model)
              │    │       │                    └ LightningGraphQA(
              │    │       │                        (model): RecursiveScriptModule(
              │    │       │                          original_name=GraphQA
              │    │       │                          (embeddings): RecursiveScriptModule(
              │    │       │                            ...
              │    │       └ <function Trainer.run_pretrain_routine at 0x7f91457f3430>
              │    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
              └ <pytorch_lightning.accelerators.gpu_backend.GPUBackend object at 0x7f910eb19940>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1239, in run_pretrain_routine
    self.train()
    │    └ <function TrainerTrainLoopMixin.train at 0x7f91457df310>
    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 394, in train
    self.run_training_epoch()
    │    └ <function TrainerTrainLoopMixin.run_training_epoch at 0x7f91457df4c0>
    └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/pytorch_lightning/trainer/training_loop.py", line 529, in run_training_epoch
    monitor_metrics = deepcopy(self.callback_metrics)
                      │        │    └ {'val_scores': {'global': score                                  tm            ...       cad          
                      │        │                            ...
                      │        └ <pytorch_lightning.trainer.trainer.Trainer object at 0x7f910ca17790>
                      └ <function deepcopy at 0x7f914bc543a0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        │      │  └ <unprintable dict object>
        │      └ {'val_scores': {'global': score                                  tm            ...       cad          
        │                              ...
        └ <function _deepcopy_dict at 0x7f914bc545e0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
    │ │        │    │        │        │      └ <unprintable dict object>
    │ │        │    │        │        └ {'global/gdtts': <Figure size 300x300 with 1 Axes>, 'global/cad': <Figure size 300x300 with 1 Axes>, 'local/lddt': <Figure si...
    │ │        │    │        └ <function deepcopy at 0x7f914bc543a0>
    │ │        │    └ <unprintable dict object>
    │ │        └ 'val_figures'
    │ └ <function deepcopy at 0x7f914bc543a0>
    └ {'val_scores': {'global': score                                  tm            ...       cad          
                            ...
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        │      │  └ <unprintable dict object>
        │      └ {'global/gdtts': <Figure size 300x300 with 1 Axes>, 'global/cad': <Figure size 300x300 with 1 Axes>, 'local/lddt': <Figure si...
        └ <function _deepcopy_dict at 0x7f914bc545e0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
    │ │        │    │        │        │      └ <unprintable dict object>
    │ │        │    │        │        └ <Figure size 300x300 with 1 Axes>
    │ │        │    │        └ <function deepcopy at 0x7f914bc543a0>
    │ │        │    └ <unprintable dict object>
    │ │        └ 'global/gdtts'
    │ └ <function deepcopy at 0x7f914bc543a0>
    └ {}
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
        │            │  │      └ (<function __newobj__ at 0x7f91a81ac790>, (<class 'matplotlib.figure.Figure'>,), {'_stale': True, 'stale_callback': None, 'fi...
        │            │  └ <unprintable dict object>
        │            └ <Figure size 300x300 with 1 Axes>
        └ <function _reconstruct at 0x7f914bc54790>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
            │        │      └ <unprintable dict object>
            │        └ {'_stale': True, 'stale_callback': None, 'figure': None, '_transform': None, '_transformSet': False, '_visible': True, '_anim...
            └ <function deepcopy at 0x7f914bc543a0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
        │      │  └ <unprintable dict object>
        │      └ {'_stale': True, 'stale_callback': None, 'figure': None, '_transform': None, '_transformSet': False, '_visible': True, '_anim...
        └ <function _deepcopy_dict at 0x7f914bc545e0>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
    │ │        │    │        │        │      └ <unprintable dict object>
    │ │        │    │        │        └ Bbox([[0.0, 0.0], [3.0, 3.0]])
    │ │        │    │        └ <function deepcopy at 0x7f914bc543a0>
    │ │        │    └ <unprintable dict object>
    │ │        └ 'bbox_inches'
    │ └ <function deepcopy at 0x7f914bc543a0>
    └ {'_stale': True, 'stale_callback': None, 'figure': None, '_transform': None, '_transformSet': False, '_visible': True, '_anim...
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/copy.py", line 153, in deepcopy
    y = copier(memo)
        │      └ <unprintable dict object>
        └ <bound method TransformNode.__copy__ of Bbox([[0.0, 0.0], [3.0, 3.0]])>
  File "/home/gabriele/miniconda3/envs/graphqa/lib/python3.8/site-packages/matplotlib/transforms.py", line 143, in __copy__
    raise NotImplementedError(

NotImplementedError: TransformNode instances can not be copied. Consider using frozen() instead.
```

What I'm missing?

**Comment by baldassarreFe** [comment]
It's an old bug in pytorch_lightning related to [this commit](https://github.com/PyTorchLightning/pytorch-lightning/commit/65b43529301143fe72dc05e45150b7fcc2aeeee6). The config says to check the validation set every 50 epochs, but the early stopping callback is run at the end of every training epoch, rather than at the end of every validation epoch. The first time the callback runs, it tries to access validation metrics that have not been computed yet and it fails.

At that time I was using a patched version of pytorch_lightning. The bug has been fixed now, but I think updating pytorch_lighning to 1.x would break other things. Thus, I recommend sticking to the version in `conda.yaml` and patching the file manually. You should be able to modify `pytorch_lightning/trainer/training_loop.py` and `pytorch_lightning/callbacks/early_stopping.py` by following the edits in the commit linked above.

Btw, I noticed that `config/train.yaml` is rather empty, refer to `runs/wandb/quirky_stallman_1133/config.yaml` that was the final configuration used.

**Comment by yuzhiguo07** [comment]
I was sticking to the version in `conda.yaml` and I found that the version of pytorch_lightning was modified. However, I still got the same error above.

**Comment by yuzhiguo07** [comment]
Hi gabrielepozzati,
I just replace the `deepcopy` by `copy` in `trainer/training_loop.py` of the `pytorch_lightning` package (version 0.9.0) and then I can run the code!
Hopefully we can get it and reproduce the results in the original paper. 
GLHF

**Comment by yuzhiguo07** [comment]
BTW, if I substitute `val/global_per_target/gdtts/R` with `global/gdtts` in `config/train.yaml` , there will be something wrong when save the checkpoint since there are no `global/gdtts` in the `metrics`. I have no idea how to make the `val/global_per_target/gdtts/R` work (I got the same error).

**Comment by baldassarreFe** [comment]
Hi, I have updated all dependencies to their latest version and changed the code as necessary. See commit b54da32. You'll notice that the callbacks are not in separate classes anymore. I liked it better before, but I had to change it because of how pytorch lightning handles the various train/val steps. 

Anyway, I was able to launch a quick training run on my laptop, so the issue with the validation metric should be fixed. Let me know if you are still having problems training on the full dataset, otherwise, I'll close this issue in a few days.
